### PR TITLE
fix shared memory allocation

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- change the interface of `allocVar` and `allocArr` (add unique id)
- add unique id to `allocArr` *example/sharedMem* code

If shared memory is allocated on `alpaka::acc::AccGpuCudaRt<...>` more than for one array or variable with the same type than all returned memory references points to the same memory address.
e.g.
```C++
...
auto& a = alpaka::block::shared::allocVar<std::uint32_t>(acc);
auto& b = alpaka::block::shared::allocVar<std::uint32_t>(acc);
assert( &a != &b ); // error both memory addresses are equal 
```

Shared memory allocations on CUDA inside a class are similar to static class variables. All classes with the same template signature has the same static variable.
```C++
template<typename T>
struct Foo {
static T mem;
};
```
`mem` is equal for all instances of Foo<T>  with the the same `T`.
The way to avoid this bug is to create a unique template signature for each in kernel shared memory allocation. Thus the problem that it is not possible in C++ to generate unique id's at compile time  the user must take care about this.

With this pull request the example looks like this:
```C++
...
auto& a = alpaka::block::shared::allocVar<std::uint32_t,0>(acc); //instead of manual counter __CONTER__ can used
auto& b = alpaka::block::shared::allocVar<std::uint32_t,1>(acc);
assert( &a != &b ); // OK!!
```

### CODE REVIEW

**Comment by erikzenker** [comment]
I ask for a solution of this problem on stackoverflow [here](http://stackoverflow.com/questions/33345498/cuda-shared-memory-wrapped-in-class-points-to-same-memory?noredirect=1#comment54487402_33345498). Maybe there is somebody out there who has some idea to solve it.

**Comment by BenjaminW3** [comment]
Thanks for reporting and 'fixing' it. I always feared that this will be an issue.
I will wait some days for answers in the SO post before merging. Maybe there is a compile-time workaround available somewhere ;-)

**Comment by psychocoderHPC** [comment]
What I found was: http://b.atch.se/posts/constexpr-counter/#solution but the problem was that it not run with nvcc? (I have not tested other compilers)

**Comment by BenjaminW3** [comment]
In #85 I have proposed a version that hides the explicit parameter and uses the trick mentioned in your link to generate compile-time IDs.

**Comment by ax3l** [comment]
ha, finally `constexpr` ;P

**Comment by psychocoderHPC** [comment]
Since another two days I tested how we can get #85  compiling on gpu. It looks like it is currently not possible.
Therefore I vote that 
 - I rebase this branch and we change the alpaka interface for shared memory allocation
 - @BenjaminW3 accept this changes

I can move the ID that it is the last parameter thus we can easily add a `unique()` if there come the day that we found a way for it.

**Comment by psychocoderHPC** [comment]
I updated this pull request and use the naming from #85

File: test/integ/sharedMem/src/main.cpp [hunk 1]
```diff
@@ -68,8 +68,8 @@ class SharedMemKernel
         std::uint32_t * const pBlockShared(acc.template getBlockSharedExternMem<std::uint32_t>());
 
         // Get some shared memory (allocate a second buffer directly afterwards to check for some synchronization bugs).
-        //std::uint32_t * const pBlockShared1(alpaka::block::shared::allocArr<std::uint32_t, TnumUselessWork::value>());
-        //std::uint32_t * const pBlockShared2(alpaka::block::shared::allocArr<std::uint32_t, TnumUselessWork::value>());
+        //std::uint32_t * const pBlockShared1(alpaka::block::shared::allocArr<__COUNTER__,std::uint32_t, TnumUselessWork::value>());
+        //std::uint32_t * const pBlockShared2(alpaka::block::shared::allocArr<__COUNTER__,std::uint32_t, TnumUselessWork::value>());
```
Comment (BenjaminW3) [inline test/integ/sharedMem/src/main.cpp]: Shouldn`t the unique id be the last template argument?

Comment (psychocoderHPC) [inline test/integ/sharedMem/src/main.cpp]: right my mistake

**Comment by psychocoderHPC** [comment]
updated

btw: I tested the new shared memory implementation on GPU (it runs)

**Pull request merged**